flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SourceStreamTask.java

```diff
@@ -147,6 +155,7 @@ protected void cancelTask() {
 
 	@Override
```

Comment:
With the above comments in mind, I would suggest to have a method in the `StreamTask` like the following:

```
@VisibleForTesting
 void finish() throws Exception {
		finished = true;
		finishTask(); // this is implemented by subclasses
	}
```
(needed in the `finishingIgnoresExceptions()` test).

And with this, the end of the `performCheckpoint` should become:

```
if (isRunning && syncSavepointLatch.isSet()) {

			final boolean checkpointWasAcked =
					syncSavepointLatch.blockUntilCheckpointIsAcknowledged();

			if (checkpointWasAcked) {
				finishTask();
			}
		}
```